plutus-use-cases/src/Plutus/Contracts/Governance.hs

```diff
@@ -0,0 +1,199 @@
+{-# LANGUAGE DataKinds          #-}
+{-# LANGUAGE DeriveAnyClass     #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts   #-}
+{-# LANGUAGE NamedFieldPuns     #-}
+{-# LANGUAGE NoImplicitPrelude  #-}
+{-# LANGUAGE RecordWildCards    #-}
+{-# LANGUAGE TemplateHaskell    #-}
+{-# LANGUAGE TypeApplications   #-}
+{-# LANGUAGE TypeFamilies       #-}
+{-# LANGUAGE TypeOperators      #-}
+{-# OPTIONS_GHC -fno-ignore-interface-pragmas #-}
+{-# OPTIONS_GHC -fno-omit-interface-pragmas #-}
+{-# OPTIONS_GHC -fno-strictness #-}
+{-# OPTIONS_GHC -fno-specialise #-}
+{-# OPTIONS_GHC -fno-spec-constr #-}
+{-# OPTIONS -fplugin-opt PlutusTx.Plugin:debug-context #-}
+module Plutus.Contracts.Governance (
+      contract
+    , scriptInstance
+    , client
+    , mkValidator
+    , votingValue
+    , Proposal(..)
+    , Voting(..)
+    , GovState(..)
+    , GovError
+    , Params(..)
+    , Schema
+    ) where
+
+import           Control.Lens                 (makeClassyPrisms, review)
+import           Control.Monad
+import           Data.Aeson                   (FromJSON, ToJSON)
+import           Data.Semigroup               (Sum (..))
+import           GHC.Generics                 (Generic)
+import           Ledger                       (MonetaryPolicyHash, PubKeyHash, Slot (..), TokenName, pubKeyHash)
+import           Ledger.Constraints           (TxConstraints)
+import qualified Ledger.Constraints           as Constraints
+import qualified Ledger.Typed.Scripts         as Scripts
+import qualified Ledger.Value                 as Value
+import           Plutus.Contract
+import           Plutus.Contract.StateMachine (AsSMContractError, State (..), StateMachine (..), Void)
+import qualified Plutus.Contract.StateMachine as SM
+import qualified PlutusTx
+import qualified PlutusTx.AssocMap            as AssocMap
+import           PlutusTx.Prelude
+import qualified Prelude
+
+data Proposal = Proposal
+    { votingDeadline :: Slot -- TODO: not used yet
+    , newLaw         :: ByteString
+    }
+    deriving stock (Show, Generic)
+    deriving anyclass (ToJSON, FromJSON)
+
+data Voting = Voting
+    { proposal :: Proposal
+    , votes    :: AssocMap.Map PubKeyHash Bool
```

Comment:
I was thinking about this over the weekend and I think there's a problem with storing the public key hashes for votes that have been cast: What prevents you from creating a new address, transfering the token to that address, and then voting a second time?

We probably need to make the voting tokens distinguishable by giving them unique `TokenName`s and then make this a map of `TokenName` to `Bool`.

Reply:
That makes a lot of sense, thanks!